pkg/reconciler/namespace/controller.go

```diff
@@ -0,0 +1,303 @@
+	c.queue.AddRateLimited(gvrstr + "::" + key)
+}
+
+func (c *Controller) enqueueNamespace(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited("namespace::" + key)
+}
+
+func (c *Controller) enqueueCluster(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.AddRateLimited("cluster::" + key)
+}
+
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer c.queue.ShutDown()
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+	}
+	klog.Infof("Starting workers")
+	<-ctx.Done()
+	klog.Infof("Stopping workers")
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	var err error
+	if strings.HasPrefix(key, "namespace::") {
+		err = c.processNamespace(ctx, strings.TrimPrefix(key, "namespace::"))
+	} else if strings.HasPrefix(key, "cluster::") {
+		err = c.processCluster(ctx, strings.TrimPrefix(key, "cluster::"))
+	} else {
+		err = c.processResource(ctx, key)
+	}
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	if _, ok := err.(*kerrors.RetryableError); ok {
+		num := c.queue.NumRequeues(key)
+		klog.Errorf("Error reconciling key %q, retrying... (#%d): %v", key, num, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	// Give up and report error elsewhere.
+	c.queue.Forget(key)
+	runtime.HandleError(err)
+	klog.Errorf("Dropping key %q after non-retryable error: %v", key, err)
+}
+
+// key is gvr::KEY
+func (c *Controller) processResource(ctx context.Context, key string) error {
+	parts := strings.SplitN(key, "::", 2)
+	if len(parts) != 2 {
+		klog.Errorf("Error parsing key %q; dropping", key)
+		return nil
+	}
+	gvrstr := parts[0]
+	gvr, _ := schema.ParseResourceArg(gvrstr)
+	if gvr == nil {
+		klog.Errorf("Error parsing GVR %q; dropping", gvrstr)
+		return nil
+	}
+	key = parts[1]
+
+	obj, exists, err := c.ddsif.IndexerFor(*gvr).GetByKey(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		klog.Infof("object %q does not exist", key)
+		return nil
+	}
+	unstr, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		klog.Errorf("object was not Unstructured: %T", obj)
+		return err
+	}
+
+	// Get logical cluster name.
+	_, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
```

Comment:
Returning this error will retry this key later, but if you can't split the key, likely that's a waste of time. Log+return nil?

Reply:
(ditto to parsing the gvr)